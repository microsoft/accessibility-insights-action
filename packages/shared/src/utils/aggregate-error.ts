// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export function throwOnAnyErrors(errors: Error[], context?: string): void {
    if (errors.length === 1) {
        throw errors[0];
    } else if (errors.length > 1) {
        throw new AggregateError(errors, context);
    }
}

export class AggregateError extends Error {
    private _errors: Error[];

    constructor(errors: Error[], context?: string) {
        const messagePrefix = (context ?? 'Multiple errors occurred') + '\n';

        const combinedSubstacks = errors.map((error) => error.stack ?? error.message).join('\n');

        const indentedCombinedSubstacks = combinedSubstacks
            .split('\n')
            .map((s) => '    ' + s)
            .join('\n');

        const tempErrorForStack = new Error(messagePrefix + indentedCombinedSubstacks);
        tempErrorForStack.name = 'AggregateError';
        const fullCombinedStack = tempErrorForStack.stack;

        const combinedSubmessages = errors.map((e) => '    ' + e.message).join('\n');
        const fullCombinedMessage = messagePrefix + combinedSubmessages;

        super(fullCombinedMessage);

        this.name = 'AggregateError';
        this.stack = fullCombinedStack;
        this._errors = errors;
    }

    get errors(): Error[] {
        return this._errors.slice();
    }
}
