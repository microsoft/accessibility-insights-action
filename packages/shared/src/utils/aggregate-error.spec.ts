// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { throwOnAnyErrors, AggregateError } from './aggregate-error';

/* eslint-disable no-regex-spaces */

describe(throwOnAnyErrors, () => {
    it('does not throw if passed zero errors', () => {
        expect(() => throwOnAnyErrors([])).not.toThrow();
    });

    it('rethrows the error as-is if passed one error', () => {
        const originalError = new Error('the original error');
        expect(() => throwOnAnyErrors([originalError])).toThrow(originalError);
    });

    it('throws an AggregateError if passed multiple errors', () => {
        const firstError = new Error('first error');
        const secondError = new Error('second error');

        try {
            throwOnAnyErrors([firstError, secondError]);
            fail(); // should throw
        } catch (e) {
            expect(e).toBeInstanceOf(AggregateError);
            expect((e as AggregateError).errors).toStrictEqual([firstError, secondError]);
        }
    });
});

describe(AggregateError, () => {
    let testSubject: AggregateError;

    beforeEach(() => {
        const firstError = new Error('first error');
        firstError.stack = `Error: first error\n    at <normalized first location>`;
        const secondError = new Error('second error');
        secondError.stack = `Error: second error\n    at <normalized second location>`;
        testSubject = new AggregateError([firstError, secondError]);
    });

    it('uses the pinned message format', () => {
        expect(testSubject.message).toMatchSnapshot();
    });

    it('uses the pinned stack format', () => {
        // We normalize the actual stack location because it depends on system paths/node version
        const stackWithLocationsCollapsed = testSubject.stack
            .replace(/^(    at) .*\n/gm, '') // This removes all stack lines except the last one (which won't have the \n)
            .replace(/^(    at) .*$/gm, '$1 <normalized location>'); // This normalizes the path/location in the last stack line

        expect(stackWithLocationsCollapsed).toMatchSnapshot();
    });

    it('uses the name "AggregateError"', () => {
        expect(testSubject.name).toBe('AggregateError');
    });

    it('is an Error', () => {
        expect(testSubject).toBeInstanceOf(Error);
    });
});
