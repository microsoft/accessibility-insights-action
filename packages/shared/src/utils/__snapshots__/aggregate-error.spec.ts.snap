// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateError uses the pinned message format 1`] = `
"Multiple errors occurred
    first error
    second error"
`;

exports[`AggregateError uses the pinned stack format 1`] = `
"AggregateError: Multiple errors occurred
    Error: first error
        at <normalized first location>
    Error: second error
        at <normalized second location>
    at <normalized location>"
`;
